### PR TITLE
[READY]Refactors techwebs to have different types of points - PR HAS NO GAMEPLAY IMPACT

### PR DESCRIPTION
**Pull request by silicons**
Still one type of point for now, this is purely a code change/refactor.

### CODE REVIEW

**Comment by arsserpentarium** [comment]
What we will do next:
Add different types of points.
Return functionality of destructive analyzer(probably as experimentor)
Make autoresearch(queue of nodes to research)
Make default queue.
Remove departmental lathes,because they are never used.
...
Honk.

**Comment by Tlaltecuhtli** [comment]
dep lathes do some weird effects on crew i late joined once and found engi lathe in sci, 1 sec and sci lathe in cargo and engi had medical for some reason

**Comment by Tlaltecuhtli** [comment]
inb4 each department has its own points gen system

**Comment by Barhandar** [comment]
Dep lathes are used. Occasionally. Either no lathes are used, someone gets the omnilathe board, or dept lathes migrate into science.

**Comment by kevinz000** [comment]
I am not removing departmental lathes .
I am not making an autoresearch queue because hopefully soon we'll get a better tech tree from me or otherwise
I am also not making research deconstruction based except for certain things like artifacts for planetstation.

File: code/modules/mob/living/carbon/human/life.dm [hunk 1]
```diff
@@ -375,10 +375,10 @@ GLOBAL_LIST_INIT(ballmer_windows_me_msg, list("Yo man, what if, we like, uh, put
 						ballmer_percent = (-abs(drunkenness - 13.35) / 0.9) + 1
 					if(prob(5))
 						say(pick(GLOB.ballmer_good_msg))
-					SSresearch.science_tech.research_points += (BALLMER_POINTS * ballmer_percent)
+					SSresearch.science_tech.add_point_type(TECHWEB_POINT_TYPE_DEFAULT, (BALLMER_POINTS * ballmer_percent))
 				if(drunkenness > 26) // by this point you're into windows ME territory
 					if(prob(5))
-						SSresearch.science_tech.research_points -= BALLMER_POINTS
+						SSresearch.science_tech.remove_point_type(TECHWEB_POINT_TYPE_DEFAULT, BALLMER_POINTS)
```
Comment (Iamgoofball) [inline code/modules/mob/living/carbon/human/life.dm]: When you expand the points, you should make this grant and remove points from all point pools.

**Comment by kevinz000** [comment]
@selea departmental lathes were part of the original plan talk with kor if you want them removed because otherwise certain designs (cough cough guns and similar) will be going by popular demand.
you can't have it both ways (a lot of fun designs, and omnilathes).
sorry, but time and time again has proven that you can't and techwebs apparently was why lots of the nerf prs were blocked, and i'd rather have certain designs in other departments than lose them forever.

**Comment by ExcessiveUseOfCobblestone** [comment]
I know it's going to be a pain but you want multiple vars rather than big list much like my nullcrate PR.

**Comment by Qustinnus** [comment]
IS THIS..FLOYDWEBS CONFIRMED???@kevinz000

**Comment by BlueNothing** [comment]
Which designs were under threat?

**Review by ExcessiveUseOfCobblestone — CHANGES_REQUESTED** [review]
multivariables vs big daddy list

**Comment by kevinz000** [comment]
@ExcessiveUseOfCobblestone i'd rather have one list for research points, as that allows me to have every type of point parsed in procs that manipulate/read points instead of having to update all of them every time a new point gets adedd
@BlueNothing primarily guns and you know as a science/sec main i can't handle not having too many weapons

**Comment by ExcessiveUseOfCobblestone** [comment]
I don't understand.

Every time points get updated generally you would have to loop through the list.

Every time points get updated specifically, you have to loop through the list even though only one matters.

**VS**

Every time points get updated generally, you make a proc for it so you're not having to go

A -= num or A.removepoints(num)
B -= num or B.removepoints(num)
C -= num or C.removepoints(num)

each instance. Adding a new one would just cause a new line in the proc.

Every time points get updated specifically, you just update the desired var. Since it's specific implementation, you'd be writing about the same (Well less since no looping) amount of code if you had desires to add new techweb currencies.

If you want to create a "misc point" list then sure, but I don't see how or even why it needs such dynamic implementation. There should not be an abundance of new techweb variants being added within a given round, nor should there be an abundance of new techweb variants after each department gets one.

**Comment by kevinz000** [comment]
False the list is keyed you don't loop through it for one type.
The key here IS dynamic implementation and ability for runtime editing.

File: code/__DEFINES/research.dm [hunk 2]
```diff
@@ -61,3 +61,21 @@
...
 
 #define RESEARCH_MATERIAL_RECLAMATION_ID "__materials"
+
+//When adding new types, update the list below!
+#define TECHWEB_POINT_TYPE_GENERIC "General Research"
+
+#define TECHWEB_POINT_TYPE_DEFAULT TECHWEB_POINT_TYPE_GENERIC
+
+//Formatted this way for readability.
+#define TECHWEB_POINT_TYPE_LIST list(\
```
Comment (SpaceManiac) [inline code/__DEFINES/research.dm]: What are these three defines for? Only one of them is even used, can it just be inlined onto `SSresearch` where `point_types` is declared?

File: code/game/objects/items.dm [hunk 3]
```diff
@@ -202,10 +202,9 @@ GLOBAL_VAR_INIT(rpg_loot_items, FALSE)
 			research_msg += sep
 			research_msg += node.display_name
 			sep = ", "
-	var/points = techweb_item_point_check(src)
-	if (points)
-		research_msg += sep
-		research_msg += "[points] points"
+	var/list/points = techweb_item_point_check(src)
+	if (length(points))
+		research_msg += techweb_point_display_generic(points)
```
Comment (SpaceManiac) [inline code/game/objects/items.dm]: Seems to be missing adding `sep` in front of the point display here.

File: code/modules/research/destructive_analyzer.dm [hunk 4]
```diff
@@ -120,19 +124,18 @@ Note: Must be placed within 3 tiles of the R&D Console
 			linked_console.stored_research.boost_with_path(SSresearch.techweb_nodes[TN.id], dpath)
 
 	else
-		var/point_value = techweb_item_point_check(loaded_item)
+		var/list/point_value = techweb_item_point_check(loaded_item)
 		if(linked_console.stored_research.deconstructed_items[loaded_item.type])
-			point_value = 0
-		var/choice = input("Are you sure you want to destroy [loaded_item] for [point_value ? "[point_value] research points" : "material reclamation"]?") in list("Proceed", "Cancel")
+			point_value = list()
+		var/choice = input("Are you sure you want to destroy [loaded_item] for [length(point_value) ? "[json_encode(point_value)]" : "material reclamation"]?") in list("Proceed", "Cancel")
```
Comment (SpaceManiac) [inline code/modules/research/destructive_analyzer.dm]: You removed the words "research points" here.

File: code/modules/research/rdconsole.dm [hunk 5]
```diff
@@ -170,7 +170,7 @@ doesn't have toxins access.
 					var/obj/item/card/id/ID = I.GetID()
 					if(istype(ID))
 						logname = "User: [ID.registered_name]"
-			stored_research.research_logs += "[logname] researched node id [id] for [price] points."
+			stored_research.research_logs += "[logname] researched node id [id] with cost [json_encode(price)]."
```
Comment (SpaceManiac) [inline code/modules/research/rdconsole.dm]: I get what you're going for but using `json_encode` for this is a little goofy, maybe change it to format nicer when you actually go to use this system.

File: code/modules/research/techweb/__techweb_helpers.dm [hunk 6]
```diff
@@ -168,16 +174,33 @@
...
+			ret += "ERRORED POINT TYPE: [pointlist[i]]"
+	return ret.Join("<BR>")
+
+/proc/techweb_point_display_rdconsole(pointlist, last_pointlist)
+	var/list/ret = list()
+	for(var/i in pointlist)
+		if(SSresearch.point_types[i])
+			ret += "[SSresearch.point_types[i]]: [pointlist[i]] (+[last_pointlist[i]])"
+		else
+			ret += "ERRORED POINT TYPE: [pointlist[i]] (+[last_pointlist[i]])"
```
Comment (SpaceManiac) [inline code/modules/research/techweb/__techweb_helpers.dm]: Can simplify this and above with something like `"[SSresearch.point_types[i] || "Error"]: ..."`

File: code/modules/research/techweb/all_nodes.dm [hunk 7]
```diff
@@ -879,14 +879,14 @@
...
-		boost_item_paths[UI.item] = 0	//allows deconning to unlock.
+		boost_item_paths |= UI.item	//allows deconning to unlock.
 
 //HELPERS
 /proc/total_techweb_exports()
 	var/list/datum/techweb_node/processing = list()
 	for(var/i in subtypesof(/datum/techweb_node))
 		processing += new i
-	. = 0
+	.
```
Comment (SpaceManiac) [inline code/modules/research/techweb/all_nodes.dm]: ?

**Review by SpaceManiac — CHANGES_REQUESTED** [review]
Can't really be confident in this system until it's been tested by actually being used, but seems mostly in order.

File: code/modules/research/rdconsole.dm (hunk 5 above)
Comment (kevinz000) [reply code/modules/research/rdconsole.dm]: @SpaceManiac if anything the input() strings are uglier, logs are generally not eye-candy-ish when viewed raw, but if you want i can try to make it better ? research_logs isn't really used rn

Comment (SpaceManiac) [reply code/modules/research/rdconsole.dm]: Not a big deal, just keep it in mind if you actually use this system.

File: code/modules/research/techweb/all_nodes.dm (hunk 7 above)
Comment (kevinz000) [reply code/modules/research/techweb/all_nodes.dm]: whoops

**Comment by kevinz000** [comment]
@SpaceManiac

**Review by SpaceManiac — CHANGES_REQUESTED** [review]
Doesn't work.

image

Suggestion: "General Research" is too long of a name. Maybe just "General".

**Comment by kevinz000** [comment]
huh

**Comment by kevinz000** [comment]
fixed @SpaceManiac

File: code/modules/research/techweb/__techweb_helpers.dm [hunk 8]
```diff
@@ -168,16 +174,30 @@
...
+		if(SSresearch.point_types[i])
+			ret += "[SSresearch.point_types[i]]: [pointlist[i]]"
+		else
+			ret += "ERRORED POINT TYPE: [pointlist[i]]"
+	return ret.Join("<BR>")
+
+/proc/techweb_point_display_rdconsole(pointlist, last_pointlist)
+	var/list/ret = list()
+	for(var/i in pointlist)
+		ret += "[SSresearch.point_types[i] || "ERRORED POINT TYPE"]: [pointlist[i]] (+[last_pointlist[i]])"
```
Comment (SpaceManiac) [inline code/modules/research/techweb/__techweb_helpers.dm]: Can you change this back to the "/ minute" form? "(+54.3)" doesn't really tell me anything and even "(+54.3/second)" is hard to reason about.

**Review by SpaceManiac — CHANGES_REQUESTED** [review]

File: code/modules/research/techweb/__techweb_helpers.dm (hunk 8 above)
Comment (kevinz000) [reply code/modules/research/techweb/__techweb_helpers.dm]: okay.

**Comment by kevinz000** [comment]
@SpaceManiac done assuming my math isn't fucked

**Comment by SpaceManiac** [comment]
Compile error.

Just a note: you don't need to rebase every time, it makes it difficult to review only the code you've changed since I've last looked.

**Comment by kevinz000** [comment]
k

**Review by SpaceManiac — APPROVED** [review]

**Comment by AlexTheSysop42** [comment]
This is making it so that you can't scan items to reveal nodes

**Comment by davethwave** [comment]
So basically your adding department specific points. Each node would require said specific kind of point?

**Pull request merged**